File: 2024/day-01/go/part1/part1.go
package main

import (
	"fmt"
	"log"
	"slices"
	"strconv"
	"strings"

	"github.com/pitoniak32/advent_of_code/2024/util-go"
)

func main() {
	inputFileContents := util.GetInputContents("day-01/input1.txt")

	result := FindDistanceSum(string(inputFileContents))

	fmt.Println(result)
}

func FindDistanceSum(input string) int {

	firstList, secondList := GetIdLists(string(input))
	fmt.Println(firstList)
	fmt.Println(secondList)
	firstLen := len(firstList)
	secondLen := len(secondList)
	if firstLen != secondLen {
		log.Fatalf("lists do not have the same length (%d != %d). Cannot continue", firstLen, secondLen)
	}
	var total = 0
	for i, _ := range firstList {
		total += util.Abs(firstList[i] - secondList[i])
	}

	return total
}

func GetIdLists(input string) ([]int, []int) {
	firstRow, secondRow := []int{}, []int{}

	lines := strings.Split(input, "\n")
	for _, line := range lines {
		parts := strings.Split(line, "   ")
		if len(parts) != 2 {
			log.Fatalf("each line should contain 2 items 3 spaces apart. Got %s", line)
		}
		fmt.Printf("l: %s, r: %s\n", parts[0], parts[1])
		first, err := strconv.Atoi(parts[0])
		if err != nil {
			log.Fatalf("first part could not be converted into integer: %v", err)
		}
		second, err := strconv.Atoi(parts[1])
		if err != nil {
			log.Fatalf("second part could not be converted into integer: %v", err)
		}
		firstRow = append(firstRow, first)
		secondRow = append(secondRow, second)
	}

	slices.Sort(firstRow)
	slices.Sort(secondRow)

	return firstRow, secondRow
}

File: 2024/day-01/rust/src/part1.rs
use anyhow::Result;

pub fn process(input: &str) -> Result<String> {
    let (mut first_list, mut second_list) = input.lines().fold(
        (Vec::<u32>::new(), Vec::<u32>::new()),
        |(mut first_list, mut second_list), line| {
            let parts = line.split("   ").collect::<Vec<_>>();
            let first = parts.first().expect("input should have a first value");
            let second = parts.get(1).expect("input should have a second value");
            first_list.push(
                first
                    .parse::<u32>()
                    .expect("input should be a valid integer"),
            );
            second_list.push(
                second
                    .parse::<u32>()
                    .expect("input should be a valid integer"),
            );
            (first_list, second_list)
        },
    );

    if first_list.len() != second_list.len() {
        panic!("the lists must be the same length")
    }

    first_list.sort();
    second_list.sort();

    let result = first_list
        .iter()
        .zip(second_list)
        .map(|(first, second)| first.abs_diff(second))
        .sum::<u32>();

    Ok(result.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;
    use pretty_assertions::assert_eq;

    #[test]
    fn test_process() -> Result<()> {
        let input = "3   4
4   3
2   5
1   3
3   9
3   3";
        assert_eq!(process(input)?, "11");
        Ok(())
    }
}
